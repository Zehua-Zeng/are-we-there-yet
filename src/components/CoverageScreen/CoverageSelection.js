import React, { useState } from 'react';
// import ReactDOM from "react-dom";
// local components
import CoverageCell from './CoverageCell';
// bootstrap components
import Button from 'react-bootstrap/Button';
// npm-installed component
// Read more here: https://www.npmjs.com/package/react-tag-autocomplete

// import misc
// styled components
import styled from 'styled-components';

const StyledCoverageSelection = styled.div`
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    .coverage-selection-list-wrapper {
        margin-right: 100px;
    }
    .coverage-selection-lable {
        font-weight: 600;
        font-size: 18;
        margin-bottom: 5px;
    }
    .coverage-selection-list {
        display: flex;
        flex-direction: column;
        
        .coverage-selection-item {
            display: flex;
            flex-direction: row;
            align-items: center;

            .tag-styled-text {
                margin: 5px;
                padding: 5px 10px;
                background-color: lightgray;
                border-radius: 5px;
                font-size: 12px;
            }
            
            i {
                padding: 0 10px 0 10px;
                color: darkgray;
                cursor: pointer;
            }

            i:hover {
                color: gray;
            }

        }
        

    }
 
    .coverage-selection-matrix-row {
        display: flex;
        min-width: 600px;
        width: 50%;

        .coverage-selection-top-lable {
            padding-top: 40px;
            font-size: 16px;
            font-weight: 600;
        }
        .coverage-selection-lable {
            font-size: 16px;
            align-self: center;
            font-weight: 600;
        }
    }
`;

const StyledCoverageContainer = styled.div`
    
    width: min-content;
    padding: 1rem 0rem;
    margin: auto;

    .matrix-1d {
        display: grid;
        grid-template-columns: repeat(12, 2.2rem);
    }

    .matrix-2d {
        display: grid;
        grid-template-columns: repeat(13, 2.2rem);
    }

    .grid-titles {
        justify-self: center;
        align-self: center;
    }
`;

const CoverageSelection = ({types, coverage_data, modPapers}) => {
    let gridStatus = {};
    // variables tracks which are selected on each row
    const [selectedFirstEncoding, setSelectedFirstEncoding] = useState("");
    const [selectedSecondEncoding, setSelectedSecondEncoding] = useState("");
    const [selectedThirdEncoding, setSelectedThirdEncoding] = useState("");
    // a list that record all selected encoding combinations
    // it's an object with 
    //    keys = sorted,concatenated encoding,
    //    values = sorted array of non-emplty selection
    const [encodingList, setEncodingList] = useState({});

    // toggle selection for encodings
    const handleClick = (num, name) => {
        if (num === 1) {
            if (selectedFirstEncoding === name) setSelectedFirstEncoding('');
            else setSelectedFirstEncoding(name);
        } else if (num === 2) {
            if (selectedSecondEncoding === name) setSelectedSecondEncoding('');
            else setSelectedSecondEncoding(name);
        } else if (num === 3) {
            if (selectedThirdEncoding === name) setSelectedThirdEncoding('');
            else setSelectedThirdEncoding(name);
        }
    };

    const addToEncodingList = (combination) => {
        // ignore unselected fields
        let arr = combination.filter( e => e !== '');
        // sort in alphabetical order
        arr.sort();
        
        // create key
        var k =  arr.join();

        // if such key doesn't exist in the list, 
        // create a new object with new k-v pair and update old list.
        if (encodingList[k] === undefined) {
            let newList = JSON.parse(JSON.stringify(encodingList));
            newList[k] = arr;
            setEncodingList(newList);
            let lst = getData(arr.length, k.replace(',',''));
            // add related papers
            modPapers(true, lst);
        } else {
            console.log(`Error: the key '${k}' already exists in the list.`);
        }
    };

    const removeFromEncodingList = (combination) => {
        // ignore unselected fields
        let arr = combination.filter( e => e !== '');
        // sort in alphabetical order
        arr.sort();
        
        // create key
        var k =  arr.join();
        // if such key exists in the list, 
        // create a new object without such k-v pair and update old list.
        if (encodingList[k] !== undefined) {
            let newList = JSON.parse(JSON.stringify(encodingList));
            delete newList[k];
            setEncodingList(newList);
            let lst = getData(arr.length, k.replace(',',''));
            // remove related papers
            modPapers(false, lst);
        } else {
            console.log(`Error: Cant remove key '${k}', it does not exist exists in the list.`);
        }
    };

    let getData = (num, key) => {
        let coverage = coverage_data[`${num}-encoding`];
        if (coverage !== undefined && Object.keys(coverage).includes(key) && coverage[key].data.length > 0) {
            return coverage[key].data;
        } else {
            return [];
        }
    };


    return (
        <StyledCoverageSelection>
            {Object.values(encodingList).length === 0 ?
                ''
                :
                <div className="coverage-selection-list-wrapper">
                    <p className="coverage-selection-lable">Selected Encodings Designs:</p>
                    <div className="coverage-selection-list">
                        {
                            Object.values(encodingList).map( item => {
                                return (
                                    <div className="coverage-selection-item">
                                        {item.map( e => (<p key={`${e}-1`} className="tag-styled-text">{e}</p>))}
                                        <i className="fas fa-times-circle" onClick={() => removeFromEncodingList(item)}></i>
                                    </div>
                                );
                            })
                        }
                        
                    </div>
                </div>
            }
            <div className="coverage-selection-select-matrix">
            <p className="coverage-selection-lable">Add an Encodings Design:</p>
                <div className="coverage-selection-matrix-row">
                    <p className="coverage-selection-top-lable">1st Encoding</p>
                    <StyledCoverageContainer>
                        <div className="matrix-1d">
                            {types.map((t) => 
                                (<div className="grid-titles" key={`title-${t}`}>{t}</div>))}
                            {types.map((t) => {
                                gridStatus[t] = false;
                                return (
                                <CoverageCell key={`${t}-2`} category='coverage-none' name={t} 
                                handleClick={() => handleClick(1,t)} isActive={selectedFirstEncoding === t} type="filter"/>);
                            })}
                        </div>
                    </StyledCoverageContainer>
                </div>
                <div className="coverage-selection-matrix-row">
                    <p className="coverage-selection-lable">2nd Encoding</p>
                    <StyledCoverageContainer>
                        <div className="matrix-1d">
                            {types.map((t) =>{
                                gridStatus[t] = false;
                                return (
                                <CoverageCell key={`${t}-3`} category='coverage-none' name={t} 
                                handleClick={() => handleClick(2,t)} isActive={selectedSecondEncoding === t} type="filter" />);
                            })}
                        </div>
                    </StyledCoverageContainer>
                </div>
                <div className="coverage-selection-matrix-row">
                    <p className="coverage-selection-lable">3rd Encoding</p>
                    <StyledCoverageContainer>
                        <div className="matrix-1d">
                            {types.map((t) =>{
                                gridStatus[t] = false;
                                return (
                                <CoverageCell key={`${t}-4`} category='coverage-none' name={t} 
                                handleClick={() => handleClick(3,t)} isActive={selectedThirdEncoding === t} type="filter" />);
                            })}
                        </div>
                    </StyledCoverageContainer>
                </div>
                <Button variant="outline-primary" block onClick={() => addToEncodingList([selectedFirstEncoding, selectedSecondEncoding, selectedThirdEncoding])}>Add</Button>
            </div>
        </StyledCoverageSelection>
    );
};

export default CoverageSelection;
