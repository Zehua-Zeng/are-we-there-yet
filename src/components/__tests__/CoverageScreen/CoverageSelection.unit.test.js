import React from 'react';
import Enzyme, { mount, shallow } from "enzyme";
import renderer from 'react-test-renderer';
import { render, screen, cleanup } from '@testing-library/react';
import CoverageSelection from '../../CoverageScreen/CoverageSelection';
// Do not remove, used in the file
import {toBeInTheDocument} from '@testing-library/jest-dom';
import Adapter from 'enzyme-adapter-react-16';
 
Enzyme.configure({ adapter: new Adapter() });

afterEach(() => {
    cleanup();
});

test('should render coverage selection as expected', () => {
    let types = [
        'PX',
        'PY',
        'L',
        'AN',
        'AR',
        'T',
        'V',
        'D',
        'S',
        'CS',
        'CH',
        'O',
    ];
    let coverage_data = {
        "1-encoding": {},
        "2-encoding": {},
        "3-encoding": {}
    };
    let modPapers = () => {};

    let selection = <CoverageSelection types={types} coverage_data={coverage_data} modPapers={modPapers}/>;
    render(selection);
    

    const selectionComponent = screen.getByTestId('coverage-selection');
    // expect it to be rendered in DOM
    expect(selectionComponent).toBeInTheDocument();

    const tree = renderer.create(selection).toJSON();
    expect(tree).toMatchSnapshot();
});

// a test failed to build successfully
// test('should not select duplicate encodings', () => {
//     let types = [
//         'PX',
//         'PY',
//         'L',
//         'AN',
//         'AR',
//         'T',
//         'V',
//         'D',
//         'S',
//         'CS',
//         'CH',
//         'O',
//     ];
//     let coverage_data = {
//         "1-encoding": {},
//         "2-encoding": {},
//         "3-encoding": {}
//     };
//     let modPapers = () => console.log(`clicked`);
    
//     // error occuered here, no dource is found
//     const selection = mount(<CoverageSelection types={types} coverage_data={coverage_data} modPapers={modPapers}/>);

//     const tree = renderer.create(selection).toJSON();
//     console.log(tree);

//     expect(selection.find('#PX-1').length).toEqual(1);
//     selection.find('#PX-1').simulate('click');
//     selection.find('#PX-2').simulate('click');
//     selection.find('button').simulate('click');
    
//     const selectionComponent = screen.getByTestId('PX-PX-');
//     expect(selectionComponent).toBeInTheDocument();
// });