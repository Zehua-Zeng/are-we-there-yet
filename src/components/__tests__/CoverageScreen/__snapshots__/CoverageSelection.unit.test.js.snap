// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render coverage selection as expected 1`] = `
<div
  className="sc-gtsrHT gCPYeG"
  data-testid="coverage-selection"
>
  
  <div
    className="coverage-selection-select-matrix"
  >
    <p
      className="coverage-selection-lable"
    >
      Add an Encodings Design:
    </p>
    <div
      className="coverage-selection-matrix-row"
    >
      <p
        className="coverage-selection-top-lable"
      >
        1st Encoding
      </p>
      <div
        className="sc-dlnjwi jYtHVp"
      >
        <div
          className="matrix-1d"
        >
          <div
            className="grid-titles"
          >
            PX
          </div>
          <div
            className="grid-titles"
          >
            PY
          </div>
          <div
            className="grid-titles"
          >
            L
          </div>
          <div
            className="grid-titles"
          >
            AN
          </div>
          <div
            className="grid-titles"
          >
            AR
          </div>
          <div
            className="grid-titles"
          >
            T
          </div>
          <div
            className="grid-titles"
          >
            V
          </div>
          <div
            className="grid-titles"
          >
            D
          </div>
          <div
            className="grid-titles"
          >
            S
          </div>
          <div
            className="grid-titles"
          >
            CS
          </div>
          <div
            className="grid-titles"
          >
            CH
          </div>
          <div
            className="grid-titles"
          >
            O
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="PX"
            id="PX-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="PY"
            id="PY-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="L"
            id="L-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="AN"
            id="AN-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="AR"
            id="AR-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="T"
            id="T-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="V"
            id="V-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="D"
            id="D-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="S"
            id="S-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="CS"
            id="CS-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="CH"
            id="CH-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="O"
            id="O-1"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="coverage-selection-matrix-row"
    >
      <p
        className="coverage-selection-lable"
      >
        2nd Encoding
      </p>
      <div
        className="sc-dlnjwi jYtHVp"
      >
        <div
          className="matrix-1d"
        >
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="PX"
            id="PX-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="PY"
            id="PY-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="L"
            id="L-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="AN"
            id="AN-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="AR"
            id="AR-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="T"
            id="T-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="V"
            id="V-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="D"
            id="D-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="S"
            id="S-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="CS"
            id="CS-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="CH"
            id="CH-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="O"
            id="O-2"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="coverage-selection-matrix-row"
    >
      <p
        className="coverage-selection-lable"
      >
        3rd Encoding
      </p>
      <div
        className="sc-dlnjwi jYtHVp"
      >
        <div
          className="matrix-1d"
        >
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="PX"
            id="PX-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="PY"
            id="PY-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="L"
            id="L-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="AN"
            id="AN-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="AR"
            id="AR-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="T"
            id="T-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="V"
            id="V-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="D"
            id="D-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="S"
            id="S-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="CS"
            id="CS-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="CH"
            id="CH-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
          <div
            className="sc-bdnxRM lmmjcq"
            data-testid="O"
            id="O-3"
          >
            <div
              className="grid-values coverage-none-deactive"
              onClick={[Function]}
            >
              <p>
                X
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
    <button
      className="btn btn-outline-primary btn-block"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      Add
    </button>
  </div>
</div>
`;
